labs/passport-adapter/README.md

```diff
@@ -0,0 +1,205 @@
+authentication system in `@loopback/authentication@2.x`.
+
+## Installation
+
+```sh
+npm i @loopback/passport-adapter --save
+```
+
+## Background
+
+`@loopback/authentication@2.x` allows users to register authentication
+strategies that implement the interface
+[`AuthenticationStrategy`](https://apidocs.strongloop.com/@loopback%2fdocs/authentication.html#AuthenticationStrategy)
+
+Since `AuthenticationStrategy` describes a strategy with different contracts
+than the passport
+[`Strategy`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/passport/index.d.ts#L79),
+and we'd like to support the existing 500+ community passport strategies, an
+**adapter class** is created in this package to convert a passport strategy to
+the one that LoopBack 4 authentication system wants.
+
+## Usage
+
+1. Create a provider for the strategy
+
+Take the basic strategy exported from
+[`passport-http`](https://github.com/jaredhanson/passport-http) as an example,
+create a provider that returns a converted basic strategy:
+
+_Note: If you are not familiar with LoopBack provider, check the documentations
+in
+[Extending LoopBack 4](https://loopback.io/doc/en/lb4/Extending-LoopBack-4.html)_
+
+```ts
+class PassportBasicAuthProvider implements Provider<AuthenticationStrategy> {
+  value(): AuthenticationStrategy {
+    // The code that returns the converted strategy
+  }
+}
+```
+
+The Provider should have two functions:
+
+- A function that takes in the verify callback function and returns a configured
+  basic strategy. To know more about the configuration, please check
+  [the configuration guide in module `passport-http`](https://github.com/jaredhanson/passport-http#usage-of-http-basic).
+
+- A function that applies the `StrategyAdapter` to the configured basic strategy
+  instance. Then in the `value()` function, you return the converted strategy.
+
+So a full implementation of the provider is:
+
+```ts
+import {BasicStrategy, BasicVerifyFunction} from 'passport-http';
+import {StrategyAdapter} from `@loopback/passport-adapter`;
+import {AuthenticationStrategy} from '@loopback/authentication';
+
+class PassportBasicAuthProvider implements Provider<AuthenticationStrategy> {
+  value(): AuthenticationStrategy {
+    const basicStrategy = this.configuredBasicStrategy(verify);
+    return this.convertToAuthStrategy(basicStrategy);
+  }
+
+  // Takes in the verify callback function and returns a configured basic strategy.
+  configuredBasicStrategy(verifyFn: BasicVerifyFunction): BasicStrategy {
+    return new BasicStrategy(verifyFn);
+  }
+
+  // Applies the `StrategyAdapter` to the configured basic strategy instance.
+  // You'd better define your strategy name as a constant, like
+  // `const AUTH_STRATEGY_NAME = 'basic'`
+  // So that you will decorate the APIs later with the same name
+  convertToAuthStrategy(basic: BasicStrategy): AuthenticationStrategy {
+    return new StrategyAdapter(basic, AUTH_STRATEGY_NAME);
+  }
+}
```

Comment:
What about the following idea?

1. Have one `PassportAuthenticationStrategy` provider to be registered with LB4 authentication strategies extension point. 

2. The `PassportAuthenticationStrategy` supports `passport:<name>`

3. Inside `PassportAuthenticationStrategy`, we expose another extension point to register all passport strategies (for example, with a tag `passport.strategy`). 
   - Creates one instance of `passport`
   - Discovers all passport strategy extensions 
   - Calls `passport.use()` to register all passport strategies

Reply:
@raymondfeng Having 2 extensions kind of falls back to the initial implementation in the draft PR https://github.com/strongloop/loopback-next/pull/2822 (which unfortunately was squashed and cannot be recovered...)

The reason why I wrapped the passport strategy directly to a common implementation of `AuthenticationStrategy` is explained in `https://github.com/strongloop/loopback-next/pull/2822/commits/dbc51ce8e07a6c4e30027dfa9146cd6a311e1243#r280376757

Reply:
And I am able to perform an authentication by invoking `passport.authenticate()`, but not quite sure about the benefit of doing it (compared with invoking `strategy.authenticate()`), and describing the type of a `passport` instance is very problematic, see [explanation](https://github.com/strongloop/loopback-next/pull/3056/files#diff-fb7bcb47eeb0d22961f9f511534a7db6R76)
By saying that I didn't mean we have to define the `passport` in any function's parameter, but in case we have to do that, we have to figure out a solution for the type.